Extract Function GUI

dpomianbuff opened the pull request:
This PR includes the following:
* Implementation of the Extract Function GUI where user can choose which function to extract
* Compute the function signature
* Candidate region adjustments bugfixes: 
   - start/end line falls in the middle of an argument list
   - start/end line falls in the middle of a parameter list
* Implementation of telemetry data collection and logging
* Some refactorings to simplify the code flow and readability

dpomianbuff committed:
ui - attempt #1

dpomianbuff committed:
Extract function UI implementation for Java code

dpomianbuff committed:
Compute function signature for Kotlin code

dpomianbuff committed:
Telemetry data collection implementation

dpomianbuff committed:
Gui refinements. Candidate region adjustment fixes

dpomianbuff committed:
Some refactorings to simplify the code flow

dpomianbuff committed:
Update readme with plugin configuration and telemetry data details

onewhl commented on README.md:
```diff
@@ -20,12 +20,128 @@ Researching collaboration between LLMs and IDEs to perform automated refactoring
+![efKotlinExample1.png](readme-images%2Fpngs%2FefKotlinExample1.png)
+
+Note that in the GUI there are two sub-panels. In the first one there's a list of all the candidates (in the example
+abovre there are two). 
```
Small typo: abo**r**ve -> above

dpomianbuff committed:
Changed plugin name. Bugfix for signature generation

dpomianbuff committed:
Region selection fix
Elapsed time spent on candidates popup telemetry data
Upgraded for IJ platform version 233

onewhl commented on README.md:
```diff
@@ -1,31 +1,170 @@
 <!-- Plugin description -->
-Researching collaboration between LLMs and IDEs to perform automated refactorings.
+Researching collaboration between LLMs and IDEs to perform automated Extract Function refactoring. It combines the 
+power of ChatGPT to provide suggestions about what fragments of code can be extracted from a large function, but leverage
```
leverage**S**?

onewhl commented on README.md:
```diff
@@ -1,31 +1,170 @@
+2. The name of the function
+3. The list of parameters
+4. The return value (if any)
+5. The body of the function is just three dots (...)
```
Could you please clarify that actual code for extraction is highlighted in the editor?

onewhl commented on README.md:
```diff
@@ -1,31 +1,170 @@
+![efKotlinExample1.png](readme-images%2Fpngs%2FefKotlinExample1.png)
+
+Note that in the GUI there are two sub-panels. In the first one there's a list of all the candidates (in the example
+abovre there are two).
```
abov**r**e -> above?

onewhl commented on src/main/kotlin/com/intellij/ml/llm/template/intentions/ApplyExtractFunctionTransformationIntention.kt:
```diff
@@ -101,57 +105,146 @@ abstract class ApplyExtractFunctionTransformationIntention(
+                it, editor, file, listOf(EFLoggerObserver(logger), candidatesApplicationTelemetryObserver)
+            )
+        }.sortedByDescending { it.lineEnd - it.lineStart }
+
```
extra lines

onewhl reviewed (approved):
@dpomianbuff thanks a lot! Please, take a look at several minor comments.

dpomianbuff committed:
Code review remarsk
